File: vs2003_cd02/Program Files/Microsoft Visual Studio .NET 2003/CompactFrameworkSDK/v1.0.5000/Windows CE/Samples/VB/Windows CE/RomanLegion/Game.vb
'
'*------------------------------------------------------------------------------
'*  <copyright from='1997' to='2002' company='Microsoft Corporation'>
'*   Copyright (c) Microsoft Corporation. All Rights Reserved.
'*
'*   This source code is intended only as a supplement to Microsoft
'*   Development Tools and/or on-line documentation.  See these other
'*   materials for detailed information regarding Microsoft code samples.
'*
'*   </copyright>
'*-------------------------------------------------------------------------------
'*
'* File: Game.vb
'*
'

Public Class GAME
    Inherits System.Windows.Forms.Form

    ' State of the game
    Private Enum Display As Integer
        Splash = 0
        Won = 1
        Game = 2
    End Enum

    ' Where we get the images
    Public m_strGamePath As String = ""

    ' Width of Game
    Private m_cxPlay As Integer

    ' Height of Game
    Private m_cyPlay As Integer

    '  Game is running ?
    Private m_bRunning As Boolean

    ' Mouse/Pen Position
    Private m_xMouse, m_yMouse As Integer

    '  Is to Repaint the game
    Private m_bRepaint As Boolean

    Private m_bmpOff As Bitmap
    Private m_gxOff As Graphics
    Private m_gxPhys As Graphics
    Private m_bmpTap() As Bitmap

    ' The Barbarian launching arrows
    Private m_Barbarian As Barbarian

    ' The Arrow
    Private m_Arrow As Arrow

    Private m_RomanArrow As Arrow

    ' Legion Attacking
    Private m_Legion As Legion

    ' Game Score
    Private m_Score As Score

    ' Last update tick
    Private m_tickOld As Integer

    ' Game status
    Private m_Status As Display

    ' Game Paused ?
    Private m_bPaused As Boolean = False

#Region " Windows Form Designer generated code "

    Public Sub New()
        MyBase.New()

        'This call is required by the Windows Form Designer.
        InitializeComponent()

        Dim strAssmPath As String = System.Reflection.Assembly.GetExecutingAssembly().GetName().CodeBase
        m_strGamePath = strAssmPath.Substring(0, strAssmPath.LastIndexOf("\") + 1)


        Me.BackColor = Color.Black

        Me.Visible = True

        ' Get game client window size
        Me.m_cxPlay = Me.ClientSize.Width
        Me.m_cyPlay = Me.ClientSize.Height

        ' Window caption
        Me.Text = "Roman Legion"

        Init()
        ' Start running
        m_bRunning = True
    End Sub

    'Form overrides dispose to clean up the component list.
    Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
        MyBase.Dispose(disposing)
    End Sub

    'NOTE: The following procedure is required by the Windows Form Designer
    'It can be modified using the Windows Form Designer.  
    'Do not modify it using the code editor.
    Private Sub InitializeComponent()
        '
        'GAME
        '
        Me.ClientSize = New System.Drawing.Size(234, 295)
        Me.MaximizeBox = False

    End Sub

    Public Shared Sub Main()
        Application.Run(New GAME())
    End Sub

#End Region


    Private Sub GameMainloop()
        Dim y As Integer
        Dim gx As Graphics

        ' get current tick count
        m_tickOld = Environment.TickCount

        InitGame()

        ' Game Running
        While (m_bRunning)

            ' Do game Step
            DoTick()

            ' Process messages to this window
            Application.DoEvents()
        End While

        ' Game over, paused or won the level
        gx = Me.CreateGraphics()

        ' Draw the bitmap correspondent to game status
        For y = 0 To Me.ClientSize.Height Step 10
            gx.DrawImage(m_bmpTap(m_Status), 0, -m_cyPlay + y)
            System.Threading.Thread.Sleep(10)
        Next

        gx.Dispose()
    End Sub

    Private Sub DoTick()
        Dim tickSleep As Integer

        ' Let the game components move
        ProcessAll()

        ' Redraw the game components
        DrawAll()

        ' Game timming (Wait next step)
        tickSleep = m_tickOld + (1000 / 70) - Environment.TickCount
        If (tickSleep > 0) Then
            System.Threading.Thread.Sleep(tickSleep)
        End If
        m_tickOld = Environment.TickCount

        RefreshAll()
        RestoreAll()
    End Sub

    Private Sub Init()

        ' Init game components
        m_Barbarian = New Barbarian(Me)
        m_Arrow = New Arrow(Me)
        m_RomanArrow = New Arrow(Me)
        m_Legion = New Legion(Me)
        m_Score = New Score(Me)

        ' Init the mouse x position as the middle of the screen
        m_xMouse = m_cxPlay / 2

        ' Load all results bitmaps
        ReDim m_bmpTap(4)

        ' The game uses the Pocket PC form size by default and only deploys
        ' the required images to save space. If you increase the form size 
        ' then the larger images are required, so ensure that the images
        ' named "WCE*.bmp" have their Build Action set to Content.

        Dim Prefix As String = "WCE"
        If (m_cxPlay <= 240) Then
            Prefix = "PPC"
        End If

        LoadBmp(Prefix + "Splash.bmp", Display.Splash)
        LoadBmp(Prefix + "Won.bmp", Display.Won)
        LoadBmp(Prefix + "Pause.bmp", Display.Game)
        LoadBmp(Prefix + "Background.bmp", 3)

        'Create offscreen playscren
        m_bmpOff = New Bitmap(m_strGamePath + Prefix + "Background.bmp")
        m_gxOff = Graphics.FromImage(m_bmpOff)
        m_gxPhys = Me.CreateGraphics()
    End Sub

    Private Sub InitGame()
        ' Game is running
        m_bRunning = True

        ' If the game was not paused
        If (Not m_bPaused) Then
            ' Disable all arrows
            m_Arrow.Stuck = True
            m_RomanArrow.Stuck = True
            m_RomanArrow.Alive = False
            ' Game paused
        Else
            ' Release the game
            m_bPaused = False
        End If

        m_bRepaint = True
    End Sub

    Private Sub ProcessAll()
        ' Get mouse position
        m_xMouse = Me.PointToClient(Control.MousePosition).X

        ' Calculate the components new position
        m_Barbarian.Process()
        m_Arrow.Process()
        m_RomanArrow.Process()
        m_Legion.Process()
    End Sub

    Private Sub DrawAll()

        ' Redraw all game components in the new position
        m_RomanArrow.Draw(m_gxOff)
        m_Arrow.Draw(m_gxOff)
        m_Barbarian.Draw(m_gxOff)
        m_Legion.Draw(m_gxOff)
        m_Score.Draw(m_gxOff)
    End Sub

    Public Sub Refresh1(ByVal rcDirty As Rectangle)
        ' Refresh components last position
        m_gxPhys.DrawImage(m_bmpOff, rcDirty.X, rcDirty.Y, rcDirty, GraphicsUnit.Pixel)
    End Sub

    Private Sub RefreshAll()

        ' Refresh the components
        Refresh1(m_Barbarian.DirtyArea)
        Refresh1(m_Arrow.DirtyArea)
        Refresh1(m_RomanArrow.DirtyArea)
        Refresh1(m_Legion.DirtyArea)
        Refresh1(m_Score.DirtyArea)

        ' Redraw everything
        If (m_bRepaint) Then
            m_gxPhys.DrawImage(m_bmpOff, 0, 0)
            m_bRepaint = False
        End If
    End Sub

    Private Sub RestoreAll()
        ' Clean where the components were
        m_Barbarian.Restore()
        m_Arrow.Restore()
        m_RomanArrow.Restore()
        m_Legion.Restore()
        m_Score.Restore()
    End Sub

    Public Sub Run()
        Application.Run(Me)
    End Sub

    Public Sub Lost()
        ' Game Over 
        ' Reset score and level
        m_Score.Value = 0
        m_Score.Level = 1

        ' Reset the Legion position
        m_Legion.Reset(m_Score.Level)

        ' Not running
        m_bRunning = False

        ' Show splash image
        m_Status = Display.Splash
    End Sub

    Public Sub Won()
        ' Game won

        ' Increase the level
        m_Score.Level = m_Score.Level + 1

        ' Reset platoon position
        m_Legion.Reset(m_Score.Level)

        ' Not running
        m_bRunning = False

        ' Show Won Image
        m_Status = Display.Won
    End Sub

    ' Expose arrow so we can start it from the legion
    Public Property RomanArrow() As Arrow
        Get
            Return m_RomanArrow
        End Get
        Set(ByVal Value As Arrow)

        End Set
    End Property

    ' Expose the Barbarian to chech game over (hit on Barbarian)
    Public Property Barbarian() As Barbarian
        Get
            Return m_Barbarian
        End Get
        Set(ByVal Value As Barbarian)

        End Set
    End Property

    ' Expose the Legion to check which roman where killed
    Public Property Legion() As Legion
        Get
            Return m_Legion
        End Get
        Set(ByVal Value As Legion)

        End Set
    End Property

    ' Expose the Score so we can update it when the Romans where killed
    Public Property Score() As Score
        Get
            Return m_Score
        End Get
        Set(ByVal Value As Score)

        End Set
    End Property


    ' Graphics Offline
    Public Property gxOff() As Graphics
        Get
            Return m_gxOff
        End Get
        Set(ByVal Value As Graphics)

        End Set
    End Property

    ' Background BMP
    Public Property BackBmp() As Bitmap
        Get
            Return m_bmpTap(3)
        End Get
        Set(ByVal Value As Bitmap)
        End Set
    End Property

    Public Property WorldBounds() As Size
        Get
            Return New Size(m_cxPlay, m_cyPlay)
        End Get
        Set(ByVal Value As Size)

        End Set
    End Property

    '  Expose the Mouse position to Barbarian get self updated
    Public Property MouseLoc() As Point
        Get
            Return New Point(m_xMouse, m_yMouse)
        End Get
        Set(ByVal Value As Point)

        End Set
    End Property

    Protected Overrides Sub OnMouseDown(ByVal mouseg As MouseEventArgs)
        Select Case m_Status
            Case Display.Game
                If (m_bRunning) Then
                    m_Barbarian.Keyboard = False

                    'Pause the Game
                    If (mouseg.Y < 15) Then
                        m_bRunning = False
                        m_bPaused = True
                    End If

                    'Release the Game
                Else
                    Me.GameMainloop()
                End If

            Case Display.Splash To Display.Won
                m_Status = Display.Game
                Me.GameMainloop()
        End Select
    End Sub

    Protected Overrides Sub OnMouseUp(ByVal mouseg As MouseEventArgs)
        Select Case m_Status
            Case Display.Game
                If (m_bRunning) Then
                    ' Launch the Arrow
                    If (m_Arrow.Stuck = True) Then
                        m_Arrow.Start(m_Barbarian.m_x + m_Barbarian.m_cx / 2, WorldBounds.Height - 37, 0, -5)
                    End If
                End If
        End Select
    End Sub

    Protected Overrides Sub OnMouseMove(ByVal mouseg As MouseEventArgs)
        ' Update mouse position
        m_xMouse = mouseg.X
        m_yMouse = mouseg.Y
    End Sub

    Protected Overrides Sub OnPaint(ByVal paintg As PaintEventArgs)
        Dim gx As Graphics
        gx = paintg.Graphics

        Select Case m_Status
            Case Display.Game
                gx.DrawImage(m_bmpTap(3), 0, 0)
                m_bRepaint = True
            Case Display.Won
                gx.DrawImage(m_bmpTap(1), 0, 0)
            Case Display.Splash
                gx.DrawImage(m_bmpTap(0), 0, 0)
        End Select
    End Sub

    Public Sub LoadBmp(ByVal bmpName As String, ByVal i As Integer)

        ' Try to Load the Bitmap
        Try
            m_bmpTap(i) = New Bitmap(m_strGamePath + bmpName)
        Catch
            ' Any error ? Create a bitmap
            m_bmpTap(i) = New Bitmap(Me.Width, Height)
            Graphics.FromImage(m_bmpTap(i)).FillRectangle(New SolidBrush(Me.BackColor), New Rectangle(0, 0, Width - 1, Height - 1))
            Graphics.FromImage(m_bmpTap(i)).DrawLine(New Pen(Color.Red), 0, 0, Width - 1, Height - 1)
            Graphics.FromImage(m_bmpTap(i)).DrawLine(New Pen(Color.Red), Width - 1, 0, 0, Height - 1)
        End Try
    End Sub

    Protected Sub RomanLegion_OnClosing(ByVal sender As Object, ByVal cancelg As System.ComponentModel.CancelEventArgs) Handles MyBase.Closing
        If (m_bRunning) Then
            m_bRunning = False
            cancelg.Cancel = True
        End If
    End Sub

    Private Sub RomanLegion_Closed(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Closed
        Me.m_bRunning = False
    End Sub

    Protected Overrides Sub OnKeyDown(ByVal keyg As KeyEventArgs)
        ' Enable to use keyboard 
        Select Case keyg.KeyData
            Case Keys.Left
                ' Start going left
                m_Barbarian.GoLeft = True
            Case Keys.Right
                ' Start going right
                m_Barbarian.GoRight = True
        End Select
    End Sub

    Protected Overrides Sub OnKeyUp(ByVal keyg As KeyEventArgs)
        Select Case keyg.KeyData
            Case Keys.Left
                ' Stop Going left
                m_Barbarian.GoLeft = False
            Case Keys.Right
                ' Stop going right
                m_Barbarian.GoRight = False
            Case Keys.Space
                ' Launch Arrow
                If (m_Arrow.Alive = False) Then
                    m_Arrow.Start(m_Barbarian.m_x + m_Barbarian.m_cx / 2, WorldBounds.Height - 37, 0, -5)
                End If
            Case Keys.P
                ' Pause game
                m_bRunning = False
                m_bPaused = True

            Case Keys.S
                ' Start game
                m_Arrow.Alive = False
                Me.GameMainloop()

        End Select
    End Sub

End Class

File: vs2003_cd02/Program Files/Microsoft Visual Studio .NET 2003/CompactFrameworkSDK/v1.0.5000/Windows CE/Samples/VC#/Windows CE/RomanLegion/Game.cs
/*=====================================================================
  File:      Game.cs

  
  ---------------------------------------------------------------------
   Copyright (C) Microsoft Corporation.  All rights reserved.

  This source code is intended only as a supplement to Microsoft
  Development Tools and/or on-line documentation.  See these other
  materials for detailed information regarding Microsoft code samples.

  THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
  PARTICULAR PURPOSE.
=====================================================================*/

using System;
using System.Drawing;
using System.Windows.Forms;
using System.Reflection;
using System.Threading;
using System.ComponentModel;

namespace RomanLegion
{
	public class GAME: Form
	{
		// State of the game
		private enum Display
		{
			Splash,
			Won,
			Game
		};
		
		// Where we get the images
		public string m_strGamePath = "";

		// Width of Game
		private int m_cxPlay;

		// Height of Game
		private int m_cyPlay;
	
		// Game is running ?
		private bool m_bRunning;
	
		// Mouse/Pen Position
		private int m_xMouse, m_yMouse;
	
		// Is to Repaint the game
		private bool m_bRepaint;
	
		private Bitmap m_bmpOff;
		private Graphics m_gxOff;
		private Graphics m_gxPhys;
		private Bitmap[] m_bmpTap;
	
		// The Barbarian launching arrows
		private Barbarian m_Barbarian;

		// The Arrow
		private Arrow m_Arrow;	

		private Arrow m_RomanArrow;	

		// Legion Attacking
	    private Legion m_Legion;

		// Game Score
		private Score m_Score;
		
		// Last update tick
		private int m_tickOld;
	
		// Game status
		private Display m_Status;

		// Game Paused ?
		private bool m_bPaused=false;
	
		public GAME()
		{

			string strAssmPath = System.Reflection.Assembly.GetExecutingAssembly().GetName().CodeBase;
			m_strGamePath = strAssmPath.Substring(0, strAssmPath.LastIndexOf(@"\") + 1);

			this.BackColor = Color.Red;

			this.Size = new Size(240, 320);

			this.Visible = true;
		
			// Get game client window size
			this.m_cxPlay = this.ClientSize.Width;
			this.m_cyPlay = this.ClientSize.Height;
		
			// Window caption
			this.Text = "Roman Legion";
		
			Init();
		
			// Start running
			m_bRunning = true;
		}

		private void GameMainloop()
		{
			int y;
			Graphics gx;
		
			// get current tick count
			m_tickOld = Environment.TickCount;
		
			InitGame();

			// Game Running
			while (m_bRunning)
			{
				// Do game Step
				DoTick();

				// Process messages to this window
				Application.DoEvents();
			}
			
			// Game over, paused or won the level
			gx = this.CreateGraphics();
			
			// Draw the bitmap correspondent to game status
			for (y = 0; y < this.ClientSize.Height; y += 10)
			{
				gx.DrawImage(m_bmpTap[(int)m_Status], 0, -m_cyPlay + y);
				Thread.Sleep(10);
			}
					
			gx.Dispose();

		}

		private void DoTick()
		{
			int tickSleep;

			// Let the game components move
			ProcessAll();

			// Redraw the game components
			DrawAll();

			// Game timming (Wait next step)
			tickSleep = m_tickOld + (1000 / 70) - Environment.TickCount;
			if (tickSleep > 0)
			{
				Thread.Sleep(tickSleep);
			}
			m_tickOld = Environment.TickCount;

			RefreshAll();
			RestoreAll();
		}

		private void Init()
		{
			// Init game components
			m_Barbarian = new Barbarian(this);
			m_Arrow = new Arrow(this);
			m_RomanArrow = new Arrow(this);
			m_Legion = new Legion(this);
			m_Score = new Score(this);
			
		
			// Init the mouse x position as the middle of the screen
			m_xMouse = m_cxPlay / 2;

			// Load all results bitmaps
			m_bmpTap = new Bitmap[4];


			// The game uses the Pocket PC form size by default and only deploys
			// the required images to save space. If you increase the form size 
			// then the larger images are required, so ensure that the images
			// named "WCE*.bmp" have their Build Action set to Content.
			string Prefix = "WCE";
			if (m_cxPlay<=240)
			{
				Prefix = "PPC";
			}

			LoadBmp(Prefix+"Splash.bmp",(int)Display.Splash);
			LoadBmp(Prefix+"Won.bmp",(int)Display.Won);
			LoadBmp(Prefix+"Pause.bmp",(int)Display.Game);
			LoadBmp(Prefix+"Background.bmp",3);
			m_bmpOff = new Bitmap(m_strGamePath+Prefix+"Background.bmp");

			// Create offscreen playscren
			
			m_gxOff = Graphics.FromImage(m_bmpOff);
			m_gxPhys = this.CreateGraphics();
		}

		private void InitGame()
		{
			// Game is running
			m_bRunning = true;
			
			// If the game was not paused
			if (!m_bPaused)
			{
				// Disable all arrows
				m_Arrow.Stuck = true;
				m_RomanArrow.Stuck = true;
				m_RomanArrow.Alive = false;
			}
			// Game paused
			else 
			{
				// Release the game
				m_bPaused = false;
			}

			m_bRepaint = true;
		}
		
		private void ProcessAll()
		{
			// Get mouse position
			m_xMouse = this.PointToClient(Control.MousePosition).X;

			// Calculate the components new position
			m_Barbarian.Process();		
			m_Arrow.Process();
			m_RomanArrow.Process();
			m_Legion.Process();
		}

		private void DrawAll()
		{
			// Redraw all game components in the new position
			m_RomanArrow.Draw(m_gxOff);
			m_Arrow.Draw(m_gxOff);
			m_Barbarian.Draw(m_gxOff);
			m_Legion.Draw(m_gxOff);
			m_Score.Draw(m_gxOff);
		}

		public void Refresh(Rectangle rcDirty)
		{
			// Refresh components last position
			m_gxPhys.DrawImage(
				m_bmpOff,
				rcDirty.X,
				rcDirty.Y,
				rcDirty,
				GraphicsUnit.Pixel);
		}

		private void RefreshAll()
		{
			// Refresh the components
			Refresh(m_Barbarian.DirtyArea);
			Refresh(m_Arrow.DirtyArea);
			Refresh(m_RomanArrow.DirtyArea);
			Refresh(m_Legion.DirtyArea);
			Refresh(m_Score.DirtyArea);

			// Redraw everything
			if (m_bRepaint)
			{
				m_gxPhys.DrawImage(m_bmpOff, 0, 0);
				m_bRepaint = false;
			}
		}

		private void RestoreAll()
		{
			// Clean where the components were
			m_Barbarian.Restore();
			m_Arrow.Restore();
			m_RomanArrow.Restore();
			m_Legion.Restore();
			m_Score.Restore();
		}
		
		public void Lost()
		{
			// Game Over 
			// Reset score and level
			m_Score.Value=0;
			m_Score.Level=1;

			// Reset the Legion position
			m_Legion.Reset(m_Score.Level);
			
			// Not running
			m_bRunning = false;


			// Show splash image
			m_Status = Display.Splash;
		}

		public void Won()
		{
			// Game won

			// Increase the level
			m_Score.Level++;

			// Reset platoon position
			m_Legion.Reset(m_Score.Level);

			// Not running
			m_bRunning = false;

			// Show Won Image
			m_Status = Display.Won;

		}
	
		// Expose arrow so we can start it from the legion
		public Arrow RomanArrow
		{
			get
			{
				return(m_RomanArrow);
			}
		}

		// Expose the Barbarian to chech game over (hit on Barbarian)
		public Barbarian Barbarian
		{
			get
			{
				return(m_Barbarian);
			}
		}

		// Expose the Legion to check which roman where killed
		public Legion Legion
		{
			get
			{
				return(m_Legion);
			}
		}

		// Expose the Score so we can update it when the Romans where killed
		public Score Score
		{
			get
			{
				return(m_Score);
			}
		}

		// Graphics Offline
		public Graphics gxOff
		{
			get
			{
				return(m_gxOff);
			}
		}

		// Background BMP
		public Bitmap BackBmp
		{
			get
			{
				return(m_bmpTap[3]);
			}
		}


		public Size WorldBounds
		{
			get
			{
				return(new Size(m_cxPlay, m_cyPlay));
			}
		}
			
        // Expose the Mouse position to Barbarian get self updated
		public Point MouseLoc
		{
			get
			{
				return(new Point(m_xMouse, m_yMouse));
			}
		}
	
		protected override void OnMouseDown(MouseEventArgs mouseg)
		{
			switch(m_Status)
			{
				case Display.Game:
					if (m_bRunning)
					{
						m_Barbarian.Keyboard = false;

						// Pause the Game
						if (mouseg.Y < 10)
						{
							m_bRunning=false;
							m_bPaused=true;
						}
					}
					// Release the Game
					else
					{
						this.GameMainloop();
					}

					break;
			    // Start the game
				case Display.Splash:
				case Display.Won:
					m_Status = Display.Game;
					this.GameMainloop();
					break;
			}
		}

		protected override void OnMouseUp(MouseEventArgs mouseg)
		{
			switch(m_Status)
			{
				case Display.Game:
					if (m_bRunning)
					{
						// Launch the Arrow
						if (m_Arrow.Stuck == true)
						{
							m_Arrow.Start(
								m_Barbarian.m_x+m_Barbarian.m_cx/2,
								WorldBounds.Height-37,
								0,
								-5);
						}
					}
					break;
			}
		}

		protected override void OnMouseMove(MouseEventArgs mouseg)
		{
			// Update mouse position
			m_xMouse = mouseg.X;
			m_yMouse = mouseg.Y;
		}

		protected override void OnPaint(PaintEventArgs paintg)
		{
			Graphics gx;
		
			gx = paintg.Graphics;
		
			switch(m_Status)
			{
				case Display.Game:
					gx.DrawImage(m_bmpTap[3], 0, 0);
					m_bRepaint = true;
					break;
				case Display.Won:
					gx.DrawImage(m_bmpTap[1], 0, 0);
					break;
				case Display.Splash:
					gx.DrawImage(m_bmpTap[0], 0, 0);
					break;
			}
		}

		public void LoadBmp(string bmpName,int i)
		{
			// Try to Load the Bitmap
			try
			{
				m_bmpTap[i] = new Bitmap(m_strGamePath+bmpName);
			}

       		// Any error ? Create a bitmap 
			catch
			{
				m_bmpTap[i] = new Bitmap(this.Width, Height);
				Graphics.FromImage(
					m_bmpTap[i]).FillRectangle(new SolidBrush(this.BackColor),
						new Rectangle(0,
						0,
						Width - 1,
						Height - 1));
				Graphics.FromImage(
					m_bmpTap[i]).DrawLine(new Pen(Color.Red), 0, 0, Width - 1, Height - 1);
				Graphics.FromImage(
					m_bmpTap[i]).DrawLine(new Pen(Color.Red), Width - 1, 0, 0, Height - 1);
			}
		}
	
		protected override void OnClosing(CancelEventArgs cancelg)
		{
			// Close Game
			if (m_bRunning)
			{
				m_bRunning = false;
				cancelg.Cancel = true;
			}
		}
	
		protected override void OnClosed(EventArgs eventg)
		{
			m_bRunning = false;
		}
	
		protected override void OnKeyDown(KeyEventArgs keyg)
		{
			// Enable to use keyboard 
			switch(keyg.KeyData)
			{
				// Start going left
				case Keys.Left:
					m_Barbarian.GoLeft = true;
					break;
			
				// Start going right
				case Keys.Right:
					m_Barbarian.GoRight = true;
					break;
			}
		}

		private void InitializeComponent()
		{
			// 
			// GAME
			// 
			this.ClientSize = new System.Drawing.Size(234, 295);
			this.MaximizeBox = false;

		}
	
		protected override void OnKeyUp(KeyEventArgs keyg)
		{
			// Enable to use keyboard 
			switch(keyg.KeyData)
			{
				// Stop Going left
				case Keys.Left:
					m_Barbarian.GoLeft = false;
					break;
			
				// Stop going right
				case Keys.Right:
					m_Barbarian.GoRight = false;
					break;

				// Launch Arrow
				case Keys.Space:
					if (m_Arrow.Alive == false)
					{
						m_Arrow.Start(m_Barbarian.m_x+m_Barbarian.m_cx/2,WorldBounds.Height-37,0,-5);
					}
					break;

				// Pause game
				case Keys.P:
					m_bRunning=false;
					m_bPaused=true;
					break;
				// Start game
				case Keys.S:
					m_Arrow.Alive = false ;
					this.GameMainloop();
					break;
 			}
		}
	}

	// Run the Application
	class MAIN
	{
		public static void Main()
		{
			Application.Run(new GAME());
		}
	}
}